apps/meteor/ee/app/livechat-enterprise/server/api/lib/tags.ts

```diff
@@ -27,11 +32,46 @@ type FindTagsByIdParams = {
 
 type FindTagsByIdResult = ILivechatTag | null;
 
-export async function findTags({ text, pagination: { offset, count, sort } }: FindTagsParams): Promise<FindTagsResult> {
+// If viewAll is true, then all tags will be returned, regardless of department
+// If viewAll is false, then all public tags will be returned, and
+//  if department is specified, then all department tags will be returned
+export async function findTags({
+	userId,
+	text,
+	department,
+	viewAll,
+	pagination: { offset, count, sort },
+}: FindTagsParams): Promise<FindTagsResult> {
+	if (!(await hasPermissionAsync(userId, 'manage-livechat-tags'))) {
+		if (viewAll) {
+			viewAll = false;
+		}
+
+		if (department) {
+			if (!(await hasAccessToDepartment(userId, department))) {
+				department = undefined;
+			}
+		}
+	}
+
 	const query = {
-		...(text && { $or: [{ name: new RegExp(escapeRegExp(text), 'i') }, { description: new RegExp(escapeRegExp(text), 'i') }] }),
+		$and: [
+			...(text ? [{ $or: [{ name: new RegExp(escapeRegExp(text), 'i') }, { description: new RegExp(escapeRegExp(text), 'i') }] }] : []),
+			...(!viewAll
+				? [
+						{
+							$or: [{ departments: { $size: 0 } }, ...(department ? [{ departments: department }] : [])],
+						},
+				  ]
+				: []),
+		],
 	};
 
+	if (!query?.$and?.length) {
+		// @ts-expect-error TS is not smart enough to understand that $and is optional here
```

Comment:
hmmm, but `$and` is not optional if you're not typing it as optional. Maybe if you change the type of `query` to be `Filter<x>` would it be optional, then you can use `DeepWritable` to remove the readonly status of each type

Reply:
Yups, I tried that but it was getting messy dealing with those types, so I went with this. But let me tried to give it another try and if you have suggestions please feel free to share :v:

Reply:
found an elegant solution to :partying_face: